proxy/src/schema.rs

```diff
@@ -44,24 +77,122 @@ impl Into<ProjectId> for IdInput {
     }
 }
 
-/// Encapsulates read paths in API.
-pub struct Query;
+/// Object representing a git branch.
+#[derive(GraphQLObject)]
+struct Branch {
+    /// Name of the branch.
+    name: String,
+}
 
-#[juniper::object(Context = Context)]
-impl Query {
-    fn apiVersion() -> &str {
-        "1.0"
-    }
+/// Object representing a git tag.
+///
+/// We still need full tag support.
+#[derive(GraphQLObject)]
+struct Tag {
+    /// Name of the tag.
+    name: String,
+}
 
-    fn projects(ctx: &Context) -> FieldResult<Vec<Project>> {
-        Ok(ctx.source.get_all_projects())
-    }
+/// Representation of a person (e.g. committer, author, signer) from a repository. Usually
+/// extracted from a signature.
+#[derive(GraphQLObject)]
+struct Person {
+    /// Name part of the commit signature commit.
+    name: String,
+    /// Email part of the commit signature commit.
+    email: String,
+    /// Reference (url/uri) to a persons avatar image.
+    avatar: String,
+}
 
-    fn project(ctx: &Context, id: IdInput) -> FieldResult<Option<Project>> {
-        Ok(ctx.source.get_project(id.into()))
+/// Representation of a code commit.
+#[derive(GraphQLObject)]
+struct Commit {
+    /// Identifier of the commit in the form of a sha1 hash. Often referred to as oid or object id.
+    sha1: String,
+    /// The author of the commit.
+    author: Person,
+    /// The summary of the commit message body.
+    summary: String,
+    /// The entire commit message body.
+    message: String,
+    /// The recorded time of the commit.
+    time: String,
+}
+
+// FIXME(xla): This should be a `std::convert::TryFrom` and needs to be addressed together with
+//             consistent error handling.
+impl From<&git2::Commit<'_>> for Commit {
+    fn from(commit: &git2::Commit) -> Self {
+        let signature = commit.author();
+
+        Self {
+            sha1: commit.id().to_string(),
+            author: Person {
+                name: signature.name().unwrap_or("invalid name").into(),
+                email: signature.email().unwrap_or("invalid email").into(),
+                avatar: "https://avatars.dicebear.com/v2/human/foo.svg".into(),
+            },
+            summary: commit.summary().unwrap_or("invalid subject").into(),
```

Comment:
I think it's called `subject` as per the [official git documentation](https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration). I would prefer sticking to that.

<img width="539" alt="Screenshot 2019-12-16 at 18 08 31" src="https://user-images.githubusercontent.com/158411/70927405-2d131b00-202f-11ea-9f6d-b242a10dd146.png">

Reply:
It gets tricky here. This template is not matching how these parts of the commit are called in other places. Most notably the reference implementation https://github.com/libgit2/libgit2/blob/d5017a14284447c289dd04cae1a310dc34aa562c/src/commit.c#L28-L39

@FintanH Also mentioned to not tie these concepts to much to any particular vcs, which would force us to look at it fresh and define our nomenclature.

Reply:
Just to nit-pick some more, the actual git source (as opposed to libgit2) uses "subject":
https://github.com/git/git/blob/master/commit.c#L530.

But good point about the abstraction though. Had a look at pijul, they don't seem to have the concept of annotating patches with messages at all.

So let's leave it as is.